### ISSUE TITLE
Add Windows RKE2 Registration Command

### ISSUE THREAD

**Issue by luthermonson** [rancher/rancher#33625]
We need to 
* auto add labels, add `--label cattle.io/os=windows` for windows 
* auto add labels, add `--label cattle.io/os=linux` for linux 
* add the install.ps1 to rancher and get the url working like the install.sh and pass that url back in the reg token
* validate that node labels work in ps1 file and the xheader is properly sent to setup the node

during testing confirm that machine label selectors work for config and try to set kubelet-args etc. and see if the config propagates to the downstream node. if it doesn't work figure out why and create a new issue for it.

**Comment by sirredbeard** [comment]
Outstanding PR https://github.com/rancher/rancher/pull/33759

**Comment by slickwarren** [comment]
on master-head (256a0eb) I ran into this issue when running the new windows registration command:
```
                                                                                                                                                                                                C:\Users\Administrator>                                                                                                 C:\Users\Administrator>curl.exe --insecure -fL https://<ip>/rke2-agent-install.ps1 -o install.ps1 && powershell.exe -ExecutionPolicy Bypass -Command "./install.ps1 -Server https://<ip> -Label 'cattle.io/os=windows' -Token <token> -Worker -CaChecksum --ca-checksum <checksum>"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13609    0 13609    0     0  13609      0 --:--:-- --:--:-- --:--:-- 45821
INFO: Using default agent configuration directory C:/etc/rancher/agent


    Directory: C:\etc\rancher


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----          8/5/2021  12:44 AM                agent
INFO: Using default agent var directory C:/etc/rancher/agent
Test-CaCheckSum : Value from https://<ip>cacerts does not look like an x509 certificate, exited with
At C:\Users\Administrator\install.ps1:411 char:9
+         Test-CaCheckSum
+         ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Test-CaCheckSum

Test-CaCheckSum : Retrieved cacerts:
At C:\Users\Administrator\install.ps1:411 char:9
+         Test-CaCheckSum
+         ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Test-CaCheckSum

-----BEGIN CERTIFICATE-----
<certificate>                                                      -
----END CERTIFICATE-----                                                                                                                                                                                                                       

C:\Users\Administrator>curl.exe --insecure -fL https:///rke2-agent-install.ps1 -o install.ps1 && powershell.exe -ExecutionPolicy Bypass -Command "./install.ps1 -Server https://<ip> -Label 'cattle.io/os=windows' -Token <token> -Worker -CaChecksum --ca-checksum <checksum>"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13609    0 13609    0     0  13609      0 --:--:-- --:--:-- --:--:-- 96517
INFO: Using default agent configuration directory C:/etc/rancher/agent
INFO: Using default agent var directory C:/etc/rancher/agent
Test-CaCheckSum : Value from https://<ip>/cacerts does not look like an x509 certificate, exited with
At C:\Users\Administrator\install.ps1:411 char:9
+         Test-CaCheckSum
+         ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Test-CaCheckSum

Test-CaCheckSum : Retrieved cacerts:
At C:\Users\Administrator\install.ps1:411 char:9
+         Test-CaCheckSum
+         ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Test-CaCheckSum

-----BEGIN CERTIFICATE-----
MIIBqDCCAU2gAwIBAgIBADAKBggqhkjOPQQDAjA7MRwwGgYDVQQKExNkeW5hbWlj
bGlzdGVuZXItb3JnMRswGQYDVQQDExJkeW5hbWljbGlzdGVuZXItY2EwHhcNMjEw
ODA0MjM0OTUzWhcNMzEwODAyMjM0OTUzWjA7MRwwGgYDVQQKExNkeW5hbWljbGlz
dGVuZXItb3JnMRswGQYDVQQDExJkeW5hbWljbGlzdGVuZXItY2EwWTATBgcqhkjO
PQIBBggqhkjOPQMBBwNCAAQP2GDxQVbwW2wFpzEBgDQt9FoiLE8zBfECopUZxMeL
5Ix9O6iuuyktUJjR9wAJ0uHhYOwgqASUVvgmA3RF6v2to0IwQDAOBgNVHQ8BAf8E
BAMCAqQwDwYDVR0TAQH/BAUwAwEB/zAdBgNVHQ4EFgQUL11gAHx4qncxz3LtDWw/
aPI2oCkwCgYIKoZIzj0EAwIDSQAwRgIhAOy/O9fh8/17UE51ElEYGMHiYXn+Igum
/AaWFXc96ItCAiEA15SFD6uMxZd0vkS3gMvsSM0OCWu5f6Ai7/qdZ5EoYoo=
-----END CERTIFICATE-----

```

**Comment by phillipsj** [comment]
I am not able to reproduce this on 2.6rc. Testing now with master-head.

**Comment by slickwarren** [comment]
reopening for the above issue (after offline discussion)

**Comment by phillipsj** [comment]
@slickwarren 

https://github.com/rancher/rancher/pull/33989
https://github.com/rancher/rancher/pull/33985
https://github.com/rancher/rancher/pull/33978
https://github.com/rancher/rancher/pull/33982

**Comment by slickwarren** [comment]
on 2.6-head (ba95a58) -- **Reopening**
ran into this error on the windows node while running the registration cmd:
```

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----          8/6/2021   5:41 PM                agent
INFO: Using default agent var directory C:/etc/rancher/agent
Test-CaCheckSum : Value from https://<IP>/cacerts does not look like an x509 certificate, exited with
At C:\Users\Administrator\install.ps1:411 char:9
+         Test-CaCheckSum
+         ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Test-CaCheckSum

Test-CaCheckSum : Retrieved cacerts:
At C:\Users\Administrator\install.ps1:411 char:9
+         Test-CaCheckSum
+         ~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Test-CaCheckSum

-----BEGIN CERTIFICATE-----
<certificate>
-----END CERTIFICATE-----
```

**Comment by phillipsj** [comment]
@slickwarren logged this issue: https://github.com/rancher/rke2/issues/1566

**Comment by phillipsj** [comment]
@slickwarren once this is merged you can retest. https://github.com/rancher/rke2/pull/1567

**Comment by slickwarren** [comment]
ok, we got past the error reported above but another was encountered, and @phillipsj knows what needs to be fixed.

 cmd was not able to run rke2.exe, but powershell seems to recognize it (but runs into other errors)
```
 install complete; you may want to run:  $env:PATH+=";C:\usr\local\bin;C:\var\lib\rancher\rke2\bin"
INFO: Enabling RKE2 agent service
rke2.exe : The term 'rke2.exe' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At C:\Users\Administrator\install.ps1:427 char:13
+             rke2.exe agent service --add
+             ~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (rke2.exe:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

(above is from cmd)

below is through powershell
```
   --version, -v  print the version
PS C:\Users\Administrator> curl.exe --insecure -fL https://<IP>/rke2-agent-install.ps1 -o install.ps1 && powershell.exe -ExecutionPolicy Bypass -Command "./install.ps1 -Server https://<IP> -Label 'cattle.io/os=windows' -Token <token> -Worker -CaChecksum <cachecksum>"
At line:1 char:85
+ ... ttps://<IP>/rke2-agent-install.ps1 -o install.ps1 && powersh ...
+                                                                ~~
The token '&&' is not a valid statement separator in this version.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : InvalidEndOfLine
```

**Comment by phillipsj** [comment]
This is ready to be tested again.

**Comment by slickwarren** [comment]
on v2.6-head (f6120e0):
It appears that the windows node starts to register with rancher, as it now appears in the `nodes` list for the custom cluster. However, while running the registration command on the windows node, the user still sees these errors when running through powershell:
```
[INFO] install complete; you may want to run:  $env:PATH+=";C:\usr\local\bin;C:\var\lib\rancher\rke2\bin"
INFO: Checking if RKE2 agent service exists
INFO: RKE2 agent service not found, enabling agent service
rke2.exe : The term 'rke2.exe' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At C:\Users\Administrator\install.ps1:426 char:13
+             rke2.exe agent service --add
+             ~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (rke2.exe:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

INFO: Starting for RKE2 agent service
Start-Service : Cannot find any service with service name 'rke2'.
At C:\Users\Administrator\install.ps1:431 char:9
+         Start-Service -Name $rke2ServiceName
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (rke2:String) [Start-Service], ServiceCommandException
    + FullyQualifiedErrorId : NoServiceFoundForGivenName,Microsoft.PowerShell.Commands.StartServiceCommand

```
After waiting about 10 minutes, the windows machines (through rancher) are still in `waiting for agent to check in and apply initial plan` state, and do not appear to show progress from there. Nodes never become active

**Comment by slickwarren** [comment]
tested on 2.6-head (4def656) and still ran into issues. After some offline discussion, it looks like there were some changes made that we are still waiting on for this to work e2e. We need an RC from the RKE2 team as well
https://github.com/rancher/kontainer-driver-metadata/pull/648
https://github.com/rancher/rke2/pull/1623
https://github.com/rancher/rke2/pull/1627

**Comment by slickwarren** [comment]
tested on v2.6.0-r3 with 1.21.3-rc6+rke2r2:
It looks like the windows registration gets further now, but the pod asking for image rancher/system-agent:v0.0.1-alpha37-suc  is in image pull backoff, and the windows node is in state waiting for agent to check in and apply initial plan when viewing through cluster management.  Upon further investigation this is a linux image trying to be pulled onto a windows node

**Comment by slickwarren** [comment]
blocked by https://github.com/rancher/rancher/issues/34282

**Comment by anupama2501** [comment]
blocked https://github.com/rancher/rancher/issues/34408

**Comment by deniseschannon** [comment]
We should use this issue to track that the registration command is working as expected. 

As part of the registration command, we need to verify that the labels are added to the kube nodes and machines. 

The machine label selector was specifically tested with https://github.com/rancher/rancher/issues/33904 and no need to test separately.

**Comment by anupama2501** [comment]
Verified on v2.6-head 2e55e84

labels `--label 'cattle.io/os=linux' ` and ` -Label 'cattle.io/os=windows' ` are present in the registration command and once the nodes are registered, labels are seen on the nodes as well.
